### PR TITLE
Add experimental column pruning rules for tileable graph optimization

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Adding an experimental implementation for tileable optimization rule, this rule works only when groupby aggregation operation follows the read CSV files, we can prune the columns that not used by the following operations when read the files.

## Related issue number

#860.

### CODE REVIEW

File: mars/optimizes/tileable_graph/tests/test_column_pruning.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import pandas as pd
+
+import mars.dataframe as md
+from mars.config import option_context
+from mars.tests.core import TestBase
+from mars.optimizes.tileable_graph.core import get_tileable_mapping
+
+
+class Test(TestBase):
+    def testGroupByPruneReadCSV(self):
```
Comment (qinxuye) [inline mars/optimizes/tileable_graph/tests/test_column_pruning.py]: We should also test 
1. two groupbyagg come after same `read_csv`.
2. `read_csv` which defines `usecols` to see if it works well.

File: mars/tiles.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def _inplace_tile(self):
     def build_graph(self, graph=None, cls=DAG, tiled=False, compose=True,
                     **build_chunk_graph_kwargs):
         tileable_graph = graph if not tiled else None
-        tileable_graph_builder = TileableGraphBuilder(graph=tileable_graph, graph_cls=cls)
+        tileable_graph_builder = get_tileable_graph_builer()(graph=tileable_graph, graph_cls=cls)
```
Comment (qinxuye) [inline mars/tiles.py]: Add an optimize option here and choose right graph builder according to the option.

File: mars/utils.py [hunk 3]
```diff
@@ -702,3 +700,13 @@ def has_unknown_shape(tiled):
     if any(np.isnan(s) for s in itertools.chain(*tiled.nsplits)):
         return True
     return False
+
+
+def get_tileable_graph_builer():
```
Comment (qinxuye) [inline mars/utils.py]: This function is not need further.

File: mars/utils.py [hunk 4]
```diff
@@ -428,8 +429,12 @@ def replace_with_fetch_or_copy(n):
                     copied.add(copy.data)
             return node_to_copy[n]
 
-        tileable_graph_builder = TileableGraphBuilder(
-            graph=graph, node_processor=replace_with_fetch_or_copy)
+        if options.optimize_tileable_graph:
```
Comment (qinxuye) [inline mars/utils.py]: `build_tileable_graph` is used on the client, shall we do the tileable graph optimization on the client? I don't think so.

File: mars/utils.py [hunk 5]
```diff
@@ -729,3 +734,37 @@ def kill_process_tree(pid, include_parent=True):
             pass
     if plasma_sock_dir:
         shutil.rmtree(plasma_sock_dir, ignore_errors=True)
+
+
+def copy_tileables(*tileables, **kwargs):
+    inputs = kwargs.pop('inputs', None)
+    copy_key = kwargs.pop('copy_key', True)
+    copy_id = kwargs.pop('copy_id', True)
+    if len(tileables) == 1:
```
Comment (qinxuye) [inline mars/utils.py]: Unify logic for single and multiple outputs, no need to separate them.

File: mars/utils.py [hunk 6]
```diff
@@ -740,31 +735,24 @@ def copy_tileables(*tileables, **kwargs):
...
+        params = t.params.copy()
+        if copy_key:
+            params['_key'] = t.key
+        if copy_id:
+            params['_id'] = t.id
+        params.update(t.extra_params)
+        kws.append(params)
+    inputs = inputs or op.inputs
+    copied = op.new_tileables(inputs, kws=kws, output_limit=len(kws))
+    if len(tileables) == 1:
```
Comment (qinxuye) [inline mars/utils.py]: How about len(tileables) > 1?

**Comment by hekaisheng** [comment]
The implementation in this PR only works in local mode, distributed support will be a little different.

**Pull request merged**